src/components/Table/hooks.tsx

```diff
@@ -0,0 +1,258 @@
+        rows = getPaginatedData(rows, state.pagination.perPage, state.pagination.page);
+      }
+
+      return {
+        ...state,
+        rows,
+        originalRows: action.data,
+      };
+    }
+
+    case "GLOBAL_FILTER": {
+      const filteredRows = action.filter(state.originalRows);
+      const selectedRowsById: { [key: number]: boolean } = {};
+      state.selectedRows.forEach((row) => {
+        selectedRowsById[row.id] = !!row.selected;
+      });
+
+      return {
+        ...state,
+        rows: filteredRows.map((row) => {
+          return selectedRowsById[row.id] ? { ...row, selected: selectedRowsById[row.id] } : { ...row };
+        }),
+        filterOn: true,
+      };
+    }
+    case "SEARCH_STRING": {
+      const stateCopySearch = { ...state };
+      stateCopySearch.rows = stateCopySearch.originalRows.filter((row) => {
+        return (
+          row.cells.filter((cell) => {
+            if (cell.value.includes(action.searchString)) {
+              return true;
+            }
+            return false;
+          }).length > 0
+        );
+      });
+      return stateCopySearch;
+    }
+    default:
+      throw new Error("Invalid reducer action");
+  }
+};
+
+export const useTable = <T extends DataType>(
+  columns: ColumnType<T>[],
+  data: T[],
+  options?: UseTableOptionsType<T>
+): UseTableReturnType<T> => {
+  const columnsWithSorting: ColumnStateType<T>[] = useMemo(
+    () =>
+      columns.map((column) => {
+        return {
+          ...column,
+          label: column.label ? column.label : column.name,
+          hidden: column.hidden ? column.hidden : false,
+          sort: column.sort,
+          sorted: {
+            on: false,
+          },
+        };
+      }),
+    [columns]
+  );
+  const columnsByName = useMemo(() => getColumnsByName(columnsWithSorting), [columnsWithSorting]);
+
+  const tableData: RowType<T>[] = useMemo(() => {
+    const sortedData = sortDataInOrder(data, columnsWithSorting);
+
+    const newData = sortedData.map((row, idx) => {
+      return {
+        id: idx,
+        selected: false,
+        hidden: false,
+        original: row,
+        cells: Object.entries(row)
+          .map(([column, value]) => {
+            return {
+              hidden: columnsByName[column].hidden,
+              field: column,
+              value,
+              render: makeRender(value, columnsByName[column].render, row),
+            };
+          })
+          .filter((cell) => !cell.hidden),
+      };
+    });
+    return newData;
+  }, [data, columnsWithSorting, columnsByName]);
+
+  const reducer = createReducer<T>();
+
+  const [state, dispatch] = useReducer(reducer, {
+    columns: columnsWithSorting,
+    columnsByName,
+    originalRows: tableData,
+    rows: tableData,
+    selectedRows: [],
+    toggleAllState: false,
+    filterOn: !!options?.filter,
+    sortColumn: null,
+    paginationEnabled: !!options?.pagination,
+    pagination: {
+      page: 1,
+      perPage: 10,
```

Comment:
We should consider making this controllable by the consumer as well.

Reply:
Ok